src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,158 @@
+package duke;
+
+import java.time.LocalDate;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.TaskList;
+import duke.task.ToDo;
+
+/**
+ * Main class for the Duke application.
+ * <p>
+ * The Duke application is a task management application that allows users to manage and track
+ * tasks such as todos, deadlines, and events. It provides a command-line interface for users
+ * to interact with.
+ * </p>
+ */
+public class Duke {
+    /** Task list to store and manage all tasks. */
+    private TaskList taskList;
+
+    /** User interface object for interacting with the user. */
+    private final Ui ui;
+
+    /** Storage object for saving and loading tasks to/from a file. */
+    private final Storage storage;
+
+    private boolean isRunning;
+
+    /**
+     * Initializes a new instance of the {@code Duke} class.
+     * This constructor creates a new {@code Storage} instance to manage saving and loading tasks.
+     * It also initializes the {@code Ui} for user interaction and sets the {@code isRunning} state to true.
+     * Additionally, it loads any saved tasks from the storage into the {@code TaskList}.
+     */
+    public Duke() {
+        this.storage = new Storage();
+        this.ui = new Ui();
+        this.isRunning = true;
+        this.taskList = new TaskList(this.storage.loadTasks());
+    }
+
+    private void run() {
+        System.out.println(this.ui.displayGreeting());
+        while (this.isRunning) {
+            String input = this.ui.readInput().strip();
+            System.out.println(this.getResponse(input));
+        }
+    }
+    /**
+     * Exits the Duke application and performs necessary cleanup operations.
+     * <p>
+     * Before exiting, this method ensures the application is currently running using an assertion.
+     * It then closes the user input interface, saves any current tasks to storage,
+     * and sets the {@code isRunning} state to false.
+     * </p>
+     *
+     * @throws DukeException If any error occurs during the exit operations,
+     *                       such as issues with saving tasks to storage.
+     */
+    private void exit() throws DukeException {
+        assert this.isRunning;
+        this.ui.closeInput();
+        this.storage.saveTasks(this.taskList.getTasks());
+        this.isRunning = false;
+    }
+
+    /**
```

Comment:
I like how detailed you are in explaining what the code for the getResponse method you have written does, providing me with a better understanding of your implementation and easily compare if your implementation is correct. I noticed you also practice this in the other methods you have implemented which is impressive!